modules/core/src/main/scala/org/tessellation/infrastructure/trust/storage/TrustStorage.scala

```diff
@@ -1,52 +1,149 @@
+              .replace(trustUpdate.trust.value.some)
           }.toMap
+          store.focus(_.trust.trust).replace(updated)
+        }
+
+      def updatePredictedTrust(trustUpdates: Map[PeerId, Double]): F[Unit] =
+        trustStoreRef.update { store =>
+          val updated = store.trust.trust ++ trustUpdates.map {
+            case (k, v) =>
+              k -> store.trust.trust
+                .getOrElse(k, TrustInfo())
+                .focus(_.predictedTrust)
+                .replace(v.some)
+          }
+          store.focus(_.trust.trust).replace(updated)
         }
 
-      def getTrust: F[Map[PeerId, TrustInfo]] = trust.get
+      def getTrust: F[TrustMap] = trustStoreRef.get.map(_.trust)
+
+      def getCurrentOrdinalTrust: F[OrdinalTrustMap] = trustStoreRef.get.map(_.currentOrdinalTrust)
+
+      def getNextOrdinalTrust: F[OrdinalTrustMap] = trustStoreRef.get.map(_.nextOrdinalTrust)
+
+      def getPublicTrust: F[PublicTrust] = getTrust.map(_.toPublicTrust)
 
-      def getPublicTrust: F[PublicTrust] = getTrust.map { trustValue =>
-        PublicTrust(trustValue.mapFilter(_.publicTrust))
+      def updatePeerPublicTrustInfo(peerId: PeerId, publicTrust: PublicTrust): F[Unit] = trustStoreRef.update { store =>
+        val updated = store.trust.peerLabels.value + (peerId -> publicTrust)
+        store.focus(_.trust.peerLabels).replace(PublicTrustMap(updated))
       }
 
-      def updatePeerPublicTrustInfo(peerId: PeerId, publicTrust: PublicTrust): F[Unit] = trust.update { t =>
-        t.updatedWith(peerId) { trustInfo =>
-          trustInfo
-            .orElse(Some(TrustInfo()))
-            .map(_.copy(peerLabels = publicTrust.labels))
+      def updateNext(ordinal: SnapshotOrdinal): F[Option[SnapshotOrdinalPublicTrust]] =
+        trustStoreRef.modify { store =>
+          val next = store.nextOrdinalTrust
+
+          if (
+            Order.gteqv(ordinal, next.ordinal) &&
+            next.trust.isEmpty &&
+            store.trust.hasTrustValues
+          ) {
+            val updated = next.focus(_.trust).replace(store.trust)
+            val ordinalPublicTrust = SnapshotOrdinalPublicTrust(
+              next.ordinal,
+              store.trust.toPublicTrust
+            )
+
+            (store.focus(_.nextOrdinalTrust).replace(updated), ordinalPublicTrust.some)
+          } else (store, none)
         }
+
+      def updateNext(
+        peerId: PeerId,
+        publicTrust: SnapshotOrdinalPublicTrust
+      ): F[Unit] = trustStoreRef.update { store =>
+        val next = store.nextOrdinalTrust
+
+        val updated =
+          if (
+            next.ordinal === publicTrust.ordinal &&
+            !next.peerLabels.value.contains(peerId) &&
+            !publicTrust.labels.isEmpty
+          ) {
+            val update = next.peerLabels.add(peerId, publicTrust.labels)
+
+            next.focus(_.peerLabels).replace(update)
+          } else next
+
+        store.focus(_.nextOrdinalTrust).replace(updated)
       }
```

Comment:
Can be written in a more declarative way by using `modify` instead of `replace`:

If you use lenses then you can focus on a nested value (see that I use `_.nextOrdinalTrust.peerLabels` in `focus`) so I can just call `focus` once.

```scala
store.focus(_.nextOrdinalTrust.peerLabels).modify { peerLabels =>
  if (
    store.nextOrdinalTrust.ordinal === publicTrust.ordinal &&
    !peerLabels.value.contains(peerId) &&
    !publicTrust.labels.isEmpty
  ) { peerLabels.add(peerId, publicTrust.labels) } else peerLabels
}
```

Reply:
Changed to

```
      def updateNext(
        peerId: PeerId,
        publicTrust: SnapshotOrdinalPublicTrust
      ): F[Unit] = trustStoreRef.update { store =>
        val next = store.nextOrdinalTrust

        store.focus(_.nextOrdinalTrust.peerLabels).modify { peerLabels =>
          if (
            next.ordinal === publicTrust.ordinal &&
            !next.peerLabels.value.contains(peerId) &&
            !publicTrust.labels.isEmpty
          ) {
            peerLabels.add(peerId, publicTrust.labels)
          } else peerLabels
        }
      }
```